src/audio/crossover/crossover.c

```diff
@@ -0,0 +1,836 @@
+		crossover_reset_state_ch(&cd->state[i]);
+}
+
+/**
+ * \brief Returns the index i such that assign_sink[i] = pipe_id.
+ *
+ * The assign_sink array in the configuration maps to the pipeline ids.
+ *
+ * \return the position at which pipe_id is found in config->assign_sink.
+ *	   -1 if not found.
+ */
+static uint8_t crossover_get_stream_index(struct sof_crossover_config *config,
+					  uint32_t pipe_id)
+{
+	int i;
+	uint32_t *assign_sink = config->assign_sink;
+
+	for (i = 0; i < config->num_sinks; i++)
+		if (assign_sink[i] == pipe_id)
+			return i;
+
+	comp_cl_err(&comp_crossover, "crossover_get_stream_index() error: couldn't find any assignment for connected pipeline %u",
+		    pipe_id);
+
+	return -1;
+}
+
+/*
+ * \brief Aligns the sinks with their respective assignments
+ *	  in the configuration.
+ *
+ * Refer to sof/src/include/sof/crossover.h for more information on assigning
+ * sinks to an output.
+ *
+ * \param[out] sinks array where the sinks are assigned
+ * \return number of sinks assigned. This number should be equal to
+ *	   config->num_sinks if no errors were found.
+ */
+static int crossover_assign_sinks(struct comp_dev *dev,
+				  struct sof_crossover_config *config,
+				  struct comp_buffer **sinks)
+{
+	struct comp_buffer *sink;
+	struct list_item *sink_list;
+	int num_sinks = 0;
+	int i;
+
+	// align sink streams with their respective configurations
+	list_for_item(sink_list, &dev->bsink_list) {
+		sink = container_of(sink_list, struct comp_buffer, source_list);
+		if (sink->sink->state == dev->state) {
+			// if no config is set, then assign the sinks in order
+			if (!config) {
+				sinks[num_sinks++] = sink;
+				continue;
+			}
+
+			i = crossover_get_stream_index(config,
+						       sink->pipeline_id);
+			if (i < 0) {
+				comp_warn(dev, "crossover_assign_sinks(), could not assign sink %i",
+					  sink->pipeline_id);
+				continue;
+			}
+
+			if (sinks[i]) {
+				comp_warn(dev, "crossover_assign_sinks(), multiple sinks from pipeline %i are assigned",
+					  sink->pipeline_id);
+				continue;
+			}
+
+			sinks[i] = sink;
+			num_sinks++;
+		}
+	}
+
+	return num_sinks;
+}
+
+/**
+ * \brief Sets the state of a single LR4 filter.
+ *
+ * An LR4 filter is built by cascading two biquads in series.
+ *
+ * \param coef struct containing the coefficients of a butterworth
+ *	       high/low pass filter.
+ * \param[out] lr4 initialized struct
+ */
+static int crossover_init_coef_lr4(struct sof_eq_iir_biquad_df2t *coef,
+				   struct iir_state_df2t  *lr4)
+{
+	/* Reuse the same coefficients for biquads so we only
+	 * store one copy of it. The processing functions will
+	 * feed the same coef array to both biquads.
+	 */
+	lr4->coef = (int32_t *)coef;
+
+	/* LR4 filters are two 2nd order filters, so only need 4 delay slots
+	 * delay[0..1] -> state for first biquad
+	 * delay[2..3] -> state for second biquad
+	 */
+	lr4->delay = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+			     sizeof(uint64_t) * 4);
```

Comment:
Maybe there should be a macro to compute size based on filter order or number of biquads?